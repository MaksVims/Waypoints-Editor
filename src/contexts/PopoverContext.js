import React, {useCallback, useContext, useState} from 'react';
import {Popover, Typography} from "@mui/material";


const PopoverContextContainer = React.createContext({})

export function usePopover() {
  return useContext(PopoverContextContainer)
}

const PopoverContext = ({children}) => {
  const [isShow, setShow] = useState(false)
  const [options, setOptions] = useState(() => ({anchorEl: null, text: ''}))

  const showPopover = useCallback((event, text) => {
    setOptions(options => (
      {...options, text, anchorEl: event.domEvent.path[0]}
    ))
    setShow(true)
  }, [setShow, setOptions])

  const hidePopover = useCallback(() => setShow(false), [setShow])

  return (
    <PopoverContextContainer.Provider value={{showPopover}}>
      <Popover
        open={isShow}
        onClose={hidePopover}
        anchorEl={options?.anchorEl}
        anchorOrigin={{
          vertical: 'top',
          horizontal: 'center',
        }}
        transformOrigin={{
          vertical: 'bottom',
          horizontal: 'center',
        }}
      >
        <Typography sx={{p: 1, fontSize: 13}}>{options?.text}</Typography>
      </Popover>
      {children}
    </PopoverContextContainer.Provider>
  );
};

export default PopoverContext;