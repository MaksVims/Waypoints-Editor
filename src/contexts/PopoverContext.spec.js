import PopoverContext, {usePopover} from "./PopoverContext";
import {render, screen} from "@testing-library/react";
import userEvent from "@testing-library/user-event";

function ConsumerComponent() {
  const {showPopover} = usePopover()
  const text = 'hello'
  const e = {
    domEvent: {
      path: []
    }
  }
  const handleClick = (event) => {
    e.domEvent.path.push(event.currentTarget)
    showPopover(e, text)
  }
  return <>
    <button id="btn" onClick={handleClick}>Click</button>
  </>
}

describe(">>> C O N T E X T --- Popover Context", function () {

  it('+++ should ConsumerComponent show default view', function () {
    render(<ConsumerComponent/>)
    expect(screen.getByRole('button')).toBeInTheDocument()
    expect(screen.queryByText('hello')).toBeNull()
  });

  it('+++ should when the user clicks on the button, the context works correctly', function () {
    const {getByRole} = render(
      <PopoverContext>
        <ConsumerComponent/>
      </PopoverContext>)

    userEvent.click(getByRole('button'))
    expect(screen.queryByText('hello')).toBeInTheDocument()
  });

  it('+++ should equal snapshot active popover context', function () {
    const result = render(
      <PopoverContext>
        <ConsumerComponent/>
      </PopoverContext>)

    userEvent.click(result.getByRole('button'))
    expect(result).toMatchSnapshot()
  });
})