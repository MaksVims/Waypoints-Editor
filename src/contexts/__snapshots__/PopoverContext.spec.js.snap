// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>> C O N T E X T --- Popover Context +++ should equal snapshot active popover context 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    style="padding-right: 1024px; overflow: hidden;"
  >
    <div
      aria-hidden="true"
    >
      <button
        id="btn"
      >
        Click
      </button>
    </div>
    <div
      class="MuiModal-root MuiPopover-root css-1i0e6f3-MuiModal-root-MuiPopover-root"
      role="presentation"
    >
      <div
        aria-hidden="true"
        class="MuiBackdrop-root MuiBackdrop-invisible css-g3hgs1-MuiBackdrop-root-MuiModal-backdrop"
        style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      />
      <div
        data-test="sentinelStart"
        tabindex="0"
      />
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation8 MuiPopover-paper css-3bmhjh-MuiPaper-root-MuiPopover-paper"
        style="opacity: 1; transform: scale(1, 1); transition: opacity 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; top: 16px; left: 16px; transform-origin: -16px -16px;"
        tabindex="-1"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-m25lzf-MuiTypography-root"
        >
          hello
        </p>
      </div>
      <div
        data-test="sentinelEnd"
        tabindex="0"
      />
    </div>
  </body>,
  "container": <div
    aria-hidden="true"
  >
    <button
      id="btn"
    >
      Click
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
